libraries/eosiolib/core/eosio/crypto_ext.hpp

```diff
@@ -48,14 +49,154 @@ namespace eosio {
          return dg;
       }
    }
-      
+
    /**
     *  @defgroup crypto Crypto
     *  @ingroup core
     *  @brief Defines API for calculating and checking hashes which
     *  require activating crypto protocol feature
     */
 
+   /**
+    * Abstracts mutable G1 and G2 points
+    *
+    *  @ingroup crypto
+    */
+   template <std::size_t Size = 32>
+   struct ec_point {
+      /**
+       * Bytes of the x coordinate
+       */
+      std::vector<char> x;
+
+      /**
+       * Bytes of the y coordinate
+       */
+      std::vector<char> y;
+
+      /**
+       * Construct a point given x and y
+       *
+       * @param x_ - The x coordinate, a vector of chars
+       * @param y_ - The y coordinate, a vector of chars
+      */
+      ec_point(std::vector<char>& x_, std::vector<char>& y_)
+      :x(x_), y(y_)
+      {
+         eosio::check( x_.size() == y_.size(), "x's size must be equal to y's" );
+         eosio::check ( x_.size() == Size, "point size must match");
+      };
+
+      /**
+       * Construct a point given a serialized point
+       *
+       * @param p - The serialized point
+       */
+      ec_point(std::vector<char>& p)
+      :x(p.data(), p.data() + p.size()/2), y(p.data() + p.size()/2, p.data() + p.size())
+      {
+         eosio::check ( p.size() == Size * 2, "point size must match");
+      };
+
+      /**
+       *  Return serialzed point containing only x and y
+       */
+      std::vector<char> serialized() const {
+         std::vector<char> x_and_y( x );
+         x_and_y.insert( x_and_y.end(), y.begin(), y.end() );
+         return x_and_y;
+      }
+   };
+
+   /**
+    * Abstracts read-only G1 and G2 points
+    *
+    *  @ingroup crypto
+    */
+   template <std::size_t Size = 32>
+   struct ec_point_view {
+      /**
+       * Pointer to the x coordinate
+       */
+      const char* x;
+
+      /**
+       * Pointer to the y coordinate
+       */
+      const char* y;
+
+      /**
+       * Number of bytes in each of x and y
+       */
+      uint32_t size;
+
+      /**
+       * Construct a point given x and y
+       *
+       * @param x_ - The x coordinate, a vector of chars
+       * @param y_ - The y coordinate, a vector of chars
+       */
+      ec_point_view(std::vector<char>& x_, std::vector<char>& y_)
```

Comment:
We should have a constructor that accepts a `ec_point` with `ec_point_view::Size == `ec_point::Size` via static assertion.